File: app/src/main/java/com/example/ruslanyussupov/popularmovies/ui/MovieGridFragment.java
package com.example.ruslanyussupov.popularmovies.ui;


import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.ruslanyussupov.popularmovies.OnMovieClickListener;
import com.example.ruslanyussupov.popularmovies.R;
import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter;
import com.example.ruslanyussupov.popularmovies.data.model.Movie;
import com.example.ruslanyussupov.popularmovies.databinding.FragmentMovieGridBinding;
import com.example.ruslanyussupov.popularmovies.utils.NetworkUtils;

import java.util.ArrayList;
import java.util.List;


public class MovieGridFragment extends Fragment {

    private static final String LOG_TAG = MovieGridFragment.class.getSimpleName();

    // Bundle keys for saving instance state
    public static final String EXTRA_MOVIE = "movie";

    private static final int MOVIE_GRID_COLUMNS = 2;

    private MovieAdapter mMovieAdapter;
    private OnMovieClickListener mMovieClickListener;
    private FragmentMovieGridBinding mBinding;
    private MovieGridViewModel mViewModel;

    public MovieGridFragment() {}

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mMovieClickListener = (OnMovieClickListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement OnMovieClickListener");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        Log.d(LOG_TAG, "onCreateView");

        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_movie_grid, container, false);

        return mBinding.getRoot();

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        Log.d(LOG_TAG, "onActivityCreated");

        // Set up MovieAdapter and GridLayoutManager to RV
        mMovieAdapter = new MovieAdapter(new ArrayList<Movie>(), mMovieClickListener);
        mBinding.rvMovies.setAdapter(mMovieAdapter);
        mBinding.rvMovies.setLayoutManager(new GridLayoutManager(getActivity(), MOVIE_GRID_COLUMNS));
        int offset = getResources().getDimensionPixelOffset(R.dimen.movie_item_offset);
        mBinding.rvMovies.addItemDecoration(new ItemDecoration(offset, offset, offset, offset));

        mViewModel = ViewModelProviders.of(getActivity()).get(MovieGridViewModel.class);

        if (NetworkUtils.hasNetworkConnection(getActivity())) {
            mViewModel.getMovies().observe(this, new Observer<List<Movie>>() {
                @Override
                public void onChanged(@Nullable List<Movie> movies) {
                    if (movies == null || movies.isEmpty()) {
                        showEmptyState();
                    } else {
                        showMovies();
                        mMovieAdapter.updateData(movies);
                    }
                }
            });
        } else {
            showNoNetworkConnectionState();
        }


    }

    @Override
    public void onDestroy() {
        Log.d(LOG_TAG, "onDestroy");
        super.onDestroy();
    }

    private void showEmptyState() {
        mBinding.loadingPb.setVisibility(View.GONE);
        mBinding.rvMovies.setVisibility(View.GONE);
        mBinding.stateTv.setVisibility(View.VISIBLE);
        mBinding.stateTv.setText(R.string.empty_state);
    }

    private void showMovies() {
        mBinding.loadingPb.setVisibility(View.GONE);
        mBinding.stateTv.setVisibility(View.GONE);
        mBinding.rvMovies.setVisibility(View.VISIBLE);
    }

    private void showNoNetworkConnectionState() {
        mBinding.loadingPb.setVisibility(View.GONE);
        mBinding.rvMovies.setVisibility(View.GONE);
        mBinding.stateTv.setVisibility(View.VISIBLE);
        mBinding.stateTv.setText(R.string.no_network_connection_state);
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/browse/MovieGridFragment.kt
package com.example.ruslanyussupov.popularmovies.browse


import androidx.lifecycle.ViewModelProviders
import android.content.Context
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import com.google.android.material.snackbar.Snackbar
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.StaggeredGridLayoutManager

import com.example.ruslanyussupov.popularmovies.R
import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter
import com.example.ruslanyussupov.popularmovies.databinding.FragmentMovieGridBinding

import timber.log.Timber

import com.example.ruslanyussupov.popularmovies.adapters.MovieAdapter.*
import com.example.ruslanyussupov.popularmovies.data.DataSource.Filter
import com.example.ruslanyussupov.popularmovies.data.Status


class MovieGridFragment : Fragment() {

    private lateinit var movieAdapter: MovieAdapter
    private lateinit var movieClickListener: OnMovieClickListener
    private lateinit var binding: FragmentMovieGridBinding
    private lateinit var viewModel: MainViewModel

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnMovieClickListener) {
            movieClickListener = context
        } else {
            throw ClassCastException("$context must implement OnMovieClickListener")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        Timber.d("onCreateView")
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_movie_grid, container, false)
        return binding.root

    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        Timber.d("onActivityCreated")

        movieAdapter = MovieAdapter(movieClickListener::onMovieClick)
        binding.rvMovies.adapter = movieAdapter
        binding.rvMovies.layoutManager = StaggeredGridLayoutManager(MOVIE_GRID_COLUMNS, StaggeredGridLayoutManager.VERTICAL)

        binding.swipeRefresh.isRefreshing = false

        viewModel = activity?.run {
            ViewModelProviders.of(this).get(MainViewModel::class.java)
        } ?: throw Exception("Invalid activity.")

        binding.swipeRefresh.setOnRefreshListener {
            if (viewModel.filter.value == Filter.FAVOURITE) {
                binding.swipeRefresh.isRefreshing = false
            } else {
                viewModel.refresh()
            }
        }

        viewModel.pagedList.observe(this, Observer {pagedList ->
            Timber.d("Paged List: ${pagedList.size}")
            movieAdapter.submitList(pagedList)
        })

        viewModel.networkState.observe(this, Observer {networkState ->
            Timber.d("Network state: ${networkState?.status}")
            when (networkState?.status) {
                Status.RUNNING -> { movieAdapter.isLoading.set(true) }
                Status.SUCCESS -> { movieAdapter.isLoading.set(false) }
                Status.FAILED -> {
                    movieAdapter.isLoading.set(false)
                    showSnackBar(networkState.msg ?: "Failed.")
                }
                else -> { movieAdapter.isLoading.set(false) }
            }
        })

        viewModel.refreshState.observe(this, Observer {refreshState ->
            Timber.d("Refresh state: ${refreshState?.status}")
            when (refreshState?.status) {
                Status.RUNNING -> { binding.swipeRefresh.isRefreshing = true }
                Status.SUCCESS -> { binding.swipeRefresh.isRefreshing = false }
                Status.FAILED -> {
                    binding.swipeRefresh.isRefreshing = false
                    showSnackBar(refreshState.msg ?: "Failed.")
                }
                else -> { binding.swipeRefresh.isRefreshing = false }
            }
        })

    }

    override fun onDestroy() {
        Timber.d("onDestroy")
        super.onDestroy()
    }

    private fun showSnackBar(message: String) {
        Snackbar.make(binding.rvMovies, message, Snackbar.LENGTH_INDEFINITE)
                .setAction("Retry") { viewModel.retry() }
                .show()
    }

    companion object {
        private const val MOVIE_GRID_COLUMNS = 3
    }

}
